pairing_api/src/pairing.rs

```diff
@@ -0,0 +1,456 @@
+use {
+    crate::{
+        uri::{parse_wc_uri, ParseError},
+        Methods,
+    },
+    chrono::Utc,
+    rand::{rngs::OsRng, Rng},
+    relay_client::{websocket::Client, MessageIdGenerator},
+    relay_rpc::{
+        domain::{MessageId, Topic},
+        rpc::{
+            params::{
+                pairing_delete::PairingDeleteRequest,
+                pairing_extend::PairingExtendRequest,
+                pairing_ping::PairingPingRequest,
+                IrnMetadata,
+                Metadata,
+                Relay,
+                RelayProtocolMetadata,
+                RequestParams,
+                ResponseParamsSuccess,
+            },
+            Payload,
+            PublishError,
+            Request,
+            Response,
+            SubscriptionError,
+            SuccessfulResponse,
+            JSON_RPC_VERSION_STR,
+        },
+    },
+    serde::{Deserialize, Serialize},
+    std::{collections::HashMap, sync::Arc, time::Duration},
+    tokio::sync::Mutex,
+    wc_common::{encrypt_and_encode, EnvelopeType},
+};
+
+/// Duration for short-term expiry (5 minutes).
+pub(crate) const EXPIRY_5_MINS: u64 = 300; // 5 mins
+/// Duration for long-term expiry (30 days).
+pub(crate) const EXPIRY_30_DAYS: u64 = 24 * 30 * 60 * 60; // 30 days
+/// The relay protocol used for WalletConnect communications.
+const RELAY_PROTOCOL: &str = "irn";
+/// The version of the WalletConnect protocol.
+const VERSION: &str = "2";
+
+/// Errors that can occur during pairing operations.
+#[derive(Debug, thiserror::Error)]
+pub enum PairingClientError {
+    #[error("Subscription error")]
+    SubscriptionError(relay_client::error::Error<SubscriptionError>),
+    #[error("Topic not found")]
+    PairingNotFound,
+    #[error("Pairing with topic already exists")]
+    PairingTopicAlreadyExists,
+    #[error("PublishError error")]
+    PingError(relay_client::error::Error<PublishError>),
+    #[error("Encode error")]
+    EncodeError(String),
+    #[error("Unexpected parameter")]
+    ParseError(ParseError),
+    #[error("Time error")]
+    TimeError(String),
+}
+
+/// Detailed information about a pairing.
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
+#[serde(rename_all = "camelCase")]
+pub struct PairingInfo {
+    pub topic: String,
+    pub relay: Relay,
+    pub peer_metadata: Option<Metadata>,
+    pub expiry: u64,
+    pub active: bool,
+    pub methods: Methods,
+}
+
+/// Represents a complete pairing including symmetric key and version.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Pairing {
+    pub sym_key: String,
+    pub version: String,
+    pub pairing: PairingInfo,
+}
+
+impl Pairing {
+    pub fn try_from_url(url: &str) -> Result<Self, PairingClientError> {
+        let parsed = parse_wc_uri(url).map_err(PairingClientError::ParseError)?;
```

Comment:
Why not use thiserror `#[from] ParseError` instead of map_err?

Reply:
done and for others too